pkg/storage/kv/kv.go

```diff
@@ -16,24 +16,160 @@ package kv
+	case LowLevelCmpEqual:
+		if exists && value == c.Value {
+			return true
+		}
+	case LowLevelCmpNotEqual:
+		if exists && value != c.Value {
+			return true
+		}
+	case LowLevelCmpLess:
+		if exists && value < c.Value {
+			return true
+		}
+	case LowLevelCmpGreater:
+		if exists && value > c.Value {
+			return true
+		}
+	case LowLevelCmpExists:
+		if exists {
+			return true
+		}
+	case LowLevelCmpNotExists:
+		if !exists {
+			return true
+		}
+	default:
+		panic("unreachable")
+	}
+	return false
+}
+
+// LowLevelTxnOp represents an operation in a LowLevelTxn's `Then` or `Else` branch and will be executed according to
+// the result of checking conditions.
+type LowLevelTxnOp struct {
+	Key    string
+	OpType LowLevelTxnOpType
+	Value  string
+	// The end key when the OpType is LowLevelOpGetRange.
+	EndKey string
+	// The limit of the keys to get when the OpType is LowLevelOpGetRange.
+	Limit int
+}
+
+// KeyValuePair represents a pair of key and value.
+type KeyValuePair struct {
+	Key   string
+	Value string
+}
+
+// LowLevelTxnResultItem represents a single result of a read operation in a LowLevelTxn.
+type LowLevelTxnResultItem struct {
+	KeyValuePairs []KeyValuePair
+}
+
+// LowLevelTxnResult represents the result of a LowLevelTxn. The results of operations in `Then` or `Else` branches
+// will be listed in `ResultItems` in the same order as the operations are added.
+// For Put or Delete operations, its corresponding result is the previous value before writing.
+type LowLevelTxnResult struct {
+	Succeeded bool
+	// The results of each operation in the `Then` branch or the `Else` branch of a transaction, depending on
+	// whether `Succeeded`. The i-th result belongs to the i-th operation added to the executed branch.
+	// * For Put or Delete operations, the result is empty.
+	// * For Get operations, the result contains a key-value pair representing the get result. In case the key
+	//   does not exist, its `KeyValuePairs` field will be empty.
+	// * For GetRange operations, the result is a list of key-value pairs containing key-value paris that are scanned.
+	ResultItems []LowLevelTxnResultItem
+}
+
+// LowLevelTxn is a low-level transaction interface. It follows the same pattern of etcd's transaction
+// API. When the backend is etcd, it simply calls etcd's equivalent APIs internally. Otherwise, the
+// behavior is simulated.
+// Considering that in different backends, the kv pairs may not have equivalent property of etcd's
+// version, create-time, etc., the abstracted LowLevelTxn interface does not support comparing on them.
+// It only supports checking the value or whether the key exists.
+// Avoid reading/writing the same key multiple times in a single transaction, otherwise the behavior
+// would be undefined.
+type LowLevelTxn interface {
+	If(conditions ...LowLevelTxnCondition) LowLevelTxn
+	Then(ops ...LowLevelTxnOp) LowLevelTxn
+	Else(ops ...LowLevelTxnOp) LowLevelTxn
+	Commit(ctx context.Context) (LowLevelTxnResult, error)
```

Comment:
Do we need ctx?

Reply:
When I'm writing the interface, consitering that ectd needs calling RPC when committing a transaction, I added it, but later I found etcd's API doesn't accept the context in its methods.
I'll remove it.